### PR TITLE
Add `make typecheck` to execute mypy in the codebase

### PR DESCRIPTION
**Pull request by marcospri**
We have quite a few type annotations on the LMS codebase at the moment but they are never checked by any tool.

This PR:

- Adds a new tox env / requirement sfile `typecheck`
- A new make target also named `typecheck`
- Configures mypy with enough ignores to pass without errors in the current code base.
- Solves/Ignores a few type issues raised by mypy
- **it doesn't** run typecheck in CI or as part of make sure.

This will allow us to gauge the complexity of maintaining the type checker happy without making it mandatory yet.

### CODE REVIEW

File: lms/db/_columns.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import sqlalchemy as sa
 
 
-def varchar_enum(enum, max_length=64, nullable=False, unique=False):
+def varchar_enum(enum, max_length=64, nullable=False, unique=False) -> sa.Column:
```
Comment (marcospri) [inline lms/db/_columns.py:4]: This allows the SQLA mypy plugin to identify these as a model's field.

File: lms/resources/_js_config/__init__.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def _h_user(self):
     def _application_instance(self):
         return self._lti_user.application_instance
 
-    def add_document_url(self, document_url):
+    def add_document_url(self, document_url) -> None:
```
Comment (marcospri) [inline lms/resources/_js_config/__init__.py:44]: Methods that have type annotations in the body but not type signatures are not checked by mypy and raised as an error. 

We have this one and a few constructors.

File: lms/services/event.py [hunk 3]
```diff
@@ -31,7 +31,7 @@ def insert_event(self, event: BaseEvent):
         self._db.add(db_event)
 
         if event.user_id:
-            for role_id in event.role_ids or [None]:
+            for role_id in event.role_ids or [None]:  # type: ignore
```
Comment (marcospri) [inline lms/services/event.py:34]: This is arguably a bit of a wart, we want to insert one per role id or just one if no roles are present.

Happy to ignore it anyway.

File: lms/session.py [hunk 4]
```diff
@@ -1,8 +1,6 @@
 """The app's Pyramid session."""
 from pyramid.session import JSONSerializer, SignedCookieSessionFactory
 
-__all__ = []
```
Comment (marcospri) [inline lms/session.py:4]: Mypy wanted a type annotations here (it doesn't know which values will be in the list).

Just deleted it.

File: lms/tasks/email_digests.py [hunk 5]
```diff
@@ -148,7 +148,7 @@ def send_instructor_email_digest(
     created_before = datetime.fromisoformat(created_before)
 
     if created_after is None:
-        created_after = created_before - timedelta(days=7)
+        created_after = created_before - timedelta(days=7)  # type: ignore
```
Comment (marcospri) [inline lms/tasks/email_digests.py:151]: Changing the type of a variable is usually frowned upon, we go from str -> datetime above.

We could do a manual `cast` but ignored these two works for now.

File: pyproject.toml [hunk 6]
```diff
@@ -138,3 +138,37 @@ score = "no"
 [tool.hdev]
 project_name = "lms"
 project_type = "application"
+
+[tool.mypy]
+python_version = 3.11
+
+disable_error_code = [
```
Comment (marcospri) [inline pyproject.toml:145]: These the errors that we have in the codebase now, nothing special about them.

If we merge a version of this the goal would be to remove them one by one.

File: tox.ini [hunk 7]
```diff
@@ -102,6 +102,7 @@ commands =
     template: python3 bin/make_template {posargs}
     bddtests: sh bin/create-db lms_bddtests
     bddtests: behave {posargs:tests/bdd/}
+    typecheck: mypy lms
```
Comment (marcospri) [inline tox.ini:105]: We could eventually change from mypy to another checker, or run more than one...

File: Makefile [hunk 8]
```diff
@@ -47,6 +47,11 @@ $(call help,make lint,"lint the code and print any warnings")
 lint: python
 	@pyenv exec tox -qe lint
 
+.PHONY: typecheck
+$(call help,make typecheck,"Type check the code and print any warnings")
+typecheck: python
+	@pyenv exec tox -qe typecheck
+
```
Comment (seanh) [inline Makefile:54]: `make template` will remove this. We'll need to either add it to the cookiecutter (if we want all our apps to have it) or move it into [the `lms.mk` file](https://github.com/hypothesis/lms/blob/main/lms.mk) (which `Makefile` imports and which the cookiecutter doesn't touch)

File: lms/services/application_instance.py [hunk 9]
```diff
@@ -322,7 +322,6 @@ def create_application_instance(  # pylint:disable=too-many-arguments
             shared_secret=secrets.token_hex(32),
             lms_url=lms_url,
             requesters_email=email,
-            created=datetime.utcnow(),
```
Comment (seanh) [inline lms/services/application_instance.py:325]: ?

File: pyproject.toml (hunk 6 above)
Comment (seanh) [reply pyproject.toml:145]: You'll have to add this to the cookiecutter or to `.cookiecutter/includes/pyproject.toml`

File: tox.ini (hunk 7 above)
Comment (seanh) [reply tox.ini:105]: You'll have to add this to the cookiecutter or to `.cookiecutter/includes/tox/commands`

**Review by seanh — APPROVED** [review]

**Pull request merged**